src/main/java/org/zalando/planb/provider/AuthorizeController.java

```diff
@@ -76,97 +130,273 @@ String showAuthorizationForm(@RequestParam(value = "realm") Optional<String> rea
+        // NOTE 1: we use "findFirst", i.e. if no parameter was passed and the client has multiple Redirect URIs
+        // configured, it will take a "random" one
+        // NOTE 2: not passing the "redirect_uri" parameter allows "snooping" the configured Redirect URI(s) for known
+        // client IDs --- client IDs should be unpredictable to prevent this
         final URI redirectUri = redirectUriParam
-                .orElseGet(() -> clientData.getRedirectUris().stream().findFirst().map(URI::create)
-                        .orElseThrow(() -> new BadRequestException("Missing redirect_uri", "invalid_request", "Missing redirect_uri")));
+                .orElseGet(() ->
+                                clientData
+                                .getRedirectUris()
+                                .stream()
+                                .findFirst()
+                                .map(URI::create)
+                                .orElseThrow(() ->
+                                                  new BadRequestException("Missing redirect_uri",
+                                                          "invalid_request", "Missing redirect_uri")));
 
         validateRedirectUri(realmName, clientId, clientData, redirectUri);
 
-        model.addAttribute("responseType", responseType);
-        model.addAttribute("realm", realmName);
-        model.addAttribute("clientId", clientId);
-        model.addAttribute("scope", scope.orElse(""));
-        model.addAttribute("state", state.orElse(""));
-        model.addAttribute("redirectUri", redirectUri.toString());
-        model.addAttribute("error", error.orElse(null));
-
-        return "login";
-    }
-
-    @RequestMapping(value = "/oauth2/authorize", method = RequestMethod.POST)
-    void authorize(
-            @RequestParam(value = "response_type", required = true) String responseType,
-            @RequestParam(value = "realm") Optional<String> realmNameParam,
-            @RequestParam(value = "client_id", required = true) String clientId,
-            @RequestParam(value = "scope") Optional<String> scope,
-            @RequestParam(value = "redirect_uri") URI redirectUri,
-            @RequestParam(value = "state") Optional<String> state,
-            @RequestParam(value = "username", required = true) String username,
-            @RequestParam(value = "password", required = true) String password,
-            @RequestHeader(value = "Host") Optional<String> hostHeader,
-            HttpServletResponse response) throws IOException, URISyntaxException, JOSEException {
+        updateModelForLogin(model, responseType, realmName, clientId, scope, state, redirectUri, error);
+
+        return LOGIN_FORM;
+    }
+
+    @RequestMapping(method = RequestMethod.POST, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseBody
+    AuthorizeResponse authorizeAsJson(
+            @RequestParam(value = RESPONSE_TYPE_SNAKECASE, required = true) final String responseType,
```

Comment:
Think I was a bit diffuse in my comment. The current code is correct, but `required=true` is not necessary, because it's default. `required=false` is also not necessary, because of the nice Java 8 Optionals. Which I always prefer over the old-fashion  `!= null` checks.